common/src/main/java/org/opensearch/ml/common/memorycontainer/MemoryStorageConfig.java

```diff
@@ -0,0 +1,272 @@
+        boolean semanticStorageEnabled = false;
+        FunctionName embeddingModelType = null;
+        String embeddingModelId = null;
+        String llmModelId = null;
+        Integer dimension = null;
+        Integer maxShortTermMemories = null;
+        Integer maxInferSize = null;
+
+        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
+        while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
+            String fieldName = parser.currentName();
+            parser.nextToken();
+
+            switch (fieldName) {
+                case MEMORY_INDEX_NAME_FIELD:
+                    memoryIndexName = parser.text();
+                    break;
+                case SEMANTIC_STORAGE_ENABLED_FIELD:
+                    semanticStorageEnabled = parser.booleanValue();
+                    break;
+                case EMBEDDING_MODEL_TYPE_FIELD:
+                    embeddingModelType = FunctionName.from(parser.text());
+                    break;
+                case EMBEDDING_MODEL_ID_FIELD:
+                    embeddingModelId = parser.text();
+                    break;
+                case LLM_MODEL_ID_FIELD:
+                    llmModelId = parser.text();
+                    break;
+                case DIMENSION_FIELD:
+                    dimension = parser.intValue();
+                    break;
+                case MAX_SHORT_TERM_MEMORIES_FIELD:
+                    maxShortTermMemories = parser.intValue();
+                    break;
+                case MAX_INFER_SIZE_FIELD:
+                    maxInferSize = parser.intValue();
+                    break;
+                default:
+                    parser.skipChildren();
+                    break;
+            }
+        }
+
+        MemoryStorageConfig config = MemoryStorageConfig
+            .builder()
+            .memoryIndexName(memoryIndexName)
+            .semanticStorageEnabled(semanticStorageEnabled)
+            .embeddingModelType(embeddingModelType)
+            .embeddingModelId(embeddingModelId)
+            .llmModelId(llmModelId)
+            .dimension(dimension)
+            .maxShortTermMemories(maxShortTermMemories != null ? maxShortTermMemories : MAX_SHORT_TERM_MEMORIES_DEFAULT_VALUE)
+            .maxInferSize(maxInferSize != null ? maxInferSize : MAX_INFER_SIZE_DEFAULT_VALUE)
+            .build();
+
+        // Note: validation is already called in the constructor
+        return config;
+    }
+
+    /**
+     * Validates the memory storage configuration based on semantic storage settings
+     * and enforces field restrictions and limits.
+     */
+    public void validate() {
+        if (!semanticStorageEnabled) {
+            // When semantic storage is disabled, only allow specific fields
+            // Clear fields that aren't allowed
+            this.embeddingModelType = null;
+            this.embeddingModelId = null;
+            this.llmModelId = null;
```

Comment:
My understanding is:

If customer want to have semantic search, then they will enable this field. And they will provide embedding model to embed the memory so that we can perform k-nn search.

The basic functionality of extracting the fact from user message and storing them as a long term memory is still in place. But embedding will not be there if `semanticStorageEnabled` is false. And during this time we will just perform basic search. 

Is this not the case? Are you thinking differently?